installation-device.nix: explain sshd usage, don't include clone-config

flokli opened the pull request:
###### Motivation for this change
The configuration used to build an installation medium isn't really
suitable to be (re)used when building a real system.
    
For example, it enables automatic root login without a password on the
console, installs sshd, but prevents it from starting on boot.
    
This can be very misleading when users try to derive their real system
from it - they should use nixos-generate-config instead, as described in
the manual.

Also update the mingetty helpLine to describe the sshd setup.
  
Closes #63576
Closes #26776




###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

flokli committed:
nixos/installation-device.nix: explain sshd usage

flokli commented:
@GrahamcOfBorg test installer

samueldr commented:
@GrahamcOfBorg eval

??!?

samueldr commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -10,10 +10,6 @@ with lib;
-      # Allow "nixos-rebuild" to work properly by providing
-      # /etc/nixos/configuration.nix.
-      ./clone-config.nix
-
```
`nixos/modules/installer/virtualbox-demo.nix` was using `installer.cloneConfigExtra` from this import.

```
~/tmp/nixpkgs/nixpkgs-PR63773 2 $ nix-instantiate --arg nixpkgs '{ outPath=./.; revCount=999999; shortRev="ofborg"; }' .
/nixos/release-combined.nix -A tested
warning: dumping very large path (> 256 MiB); this may run out of memory
error: The option `installer' defined in `/Users/samuel/tmp/nixpkgs/nixpkgs-PR63773/nixos/modules/installer/virtualbox-demo.nix'
does not exist.
(use '--show-trace' to show detailed location information)
```

In addition, the cloneConfig system for installer isos is desired. Let's say I need to rebuild with a specific setting for doing something, there is currently a loss of function.

I think, instead, `installer.cloneConfig` should be set to `false` on image types where the installation media is also the installation target.

flokli replied:
I pushed a new version, which sets `installer.cloneConfig = false;` instead.

This only seemed to be necessary for `sd-image.nix`.

flokli committed:
sd-image.nix: set installer.cloneConfig to false

As SD Card images are both installation media and installation target,
don't copy over a /etc/nixos/configuration.nix

Closes #63576.

flokli committed:
nixos/doc: update installation-device.nix

This got quite outdated. We now have both the nixos and root user
available for example.

worldofpeace commented:
Oh, missed the `profiles` documentation because I don't think I noticed when this was added.

worldofpeace commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -55,13 +55,13 @@ with lib;
+    services.mingetty.helpLine = ''
+      The "nixos" and "root" accounts have empty passwords.
+      sshd can be started manually via 'systemctl start sshd'
+      - remember to set a password, or add to ~/.ssh/authorized_keys first.
```
How about
```
The "nixos" and "root" accounts have empty passwords.

Type `sudo systemctl start sshd` to the SSH daemon.
You then must set a password for either "root"  or "nixos" with `passwd` to be able to login.
```

That's inline to what's in the manual, but we don't have an entry there about `authorized_keys`.
Also it may become confusing as to which `authorized_keys` file to add to since it could either be `root`'s home `nixos`'s.

worldofpeace commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -55,13 +55,13 @@ with lib;
+      sshd can be started manually via 'systemctl start sshd'
+      - remember to set a password, or add to ~/.ssh/authorized_keys first.
+    '' + optionalString config.services.xserver.enable ''
+          Type `systemctl start display-manager' to\nstart the graphical user interface.
```
```suggestion
          Type `sudo systemctl start display-manager' to\nstart the graphical user interface.
```

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
-  A message is shown to the user to start a display manager if needed, ssh with
-  <xref linkend="opt-services.openssh.permitRootLogin"/> are enabled (but
-  doesn't autostart). WPA Supplicant is also enabled without autostart.
+  A message is shown to the user that he can log in either as
```
```suggestion
  A message is shown to the user that they can log in either as
```

flokli replied:
I went with

> A message is shown to the user to log in either as […]

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
-  <filename>/etc/nixos/configuration.nix</filename> exists (for
-  <command>nixos-rebuild</command> to work), a copy of the Nixpkgs channel
-  snapshot used to create the install media.
+  Provides a basic configuration for installation devices like CDs.
```
At this point `installation device` as a CD even feels antiquated :smile: .
Though I think renaming that is out of scope.

flokli replied:
Haha yes, the code is also in the `nixos/modules/installer/cd-dvd/` subfolder (and contains `sd-image` and `system-tarball` too) - But I guess cleaning this up should be out of scope for this PR at least.

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
+  Provides a basic configuration for installation devices like CDs.
+  This enables redistributable firmware, includes the
+  <link linkend="sec-profile-clone-config">Clone Config profile</link>
+  and a copy of the Nixpkgs channel, so nixos-install works out of the box.
```
```suggestion
  and a copy of the Nixpkgs channel, so <command>nixos-install</command> works out of the box.
```

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
-  Additionally, documentation for <link linkend="opt-documentation.enable">
-  Nixpkgs</link> and <link linkend="opt-documentation.nixos.enable">NixOS
-  </link> are forcefully enabled (to override the
+  Documentation for <link linkend="opt-documentation.enable"> Nixpkgs</link>
```
```suggestion
  Documentation for <link linkend="opt-documentation.enable">Nixpkgs</link>
```
space needed?

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
-  Autologin is enabled as root.
+  NixOS manual is shown automatically on TTY 8, udisks is disabled.
+  Autologin is enabled as <literal>nixos</literal> user, while passwordless
+  <literal>sudo</literal> is possible.
```
Is it mentioned anywhere that "root" hasn't a password?

flokli replied:
yes, but a tad late. I reshuffled things a bit, might look better now.

worldofpeace commented on nixos/doc/manual/configuration/profiles/installation-device.xml:
```diff
@@ -6,33 +6,31 @@
-  made more silent for remote public IP installs, and several settings are
-  tweaked so that the installer has a better chance of succeeding under
-  low-memory environments.
+  Several settings are tweaked so that the installer has a better chance of
```
I'm not sure if we should imply for success but to just improve it's performance in that environment.

```
Several settings are tweaked to improve Nix and the installers performance under memory-constrained environments.
```

flokli replied:
This could be misinterpreted at "making it faster", but it seems more to be a "getting it to work at all".

flokli committed:
nixos/doc/clone-config.xml: add line about installation media = installation target

worldofpeace commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -55,13 +55,15 @@ with lib;
+    services.mingetty.helpLine = ''
+      The "nixos" and "root" accounts have empty passwords.
+
+      Type `sudo systemctl start sshd` to the SSH daemon.
```
```suggestion
      Type `sudo systemctl start sshd` to start the SSH daemon.
```

flokli replied:
fixed, thanks!

worldofpeace commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -55,13 +55,15 @@ with lib;
+      You then must set a password for either "root" or "nixos"
+      with `passwd` to be able to login.
+    '' + optionalString config.services.xserver.enable ''
+          Type `sudo systemctl start display-manager' to\nstart the graphical user interface.
```
The indentation here looks a bit larger than standard.

flokli replied:
fixed, thanks!

worldofpeace reviewed (approved):
Docs seem good :+1:

flokli commented:
@samueldr, could you give this a try once `sd_image` builds again?

worldofpeace commented:
Probably immediately easy enough to revert https://github.com/NixOS/nixpkgs/pull/63835 for the image

flokli commented:
Rebased on top of master, as https://github.com/NixOS/nixpkgs/pull/66850 landed.
